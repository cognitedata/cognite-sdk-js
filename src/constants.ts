// Copyright 2019 Cognite AS

/** @hidden */
export const API_VERSION: string = 'v1';
/** @hidden */
export const BASE_URL: string = 'https://api.cognitedata.com';

/** @hidden */
export const AUTHORIZATION_HEADER = 'Authorization';
/** @hidden */
export const API_KEY_HEADER = 'api-key';
/** @hidden */
export const X_CDF_APP_HEADER = 'x-cdp-app';
/** @hidden */
export const X_CDF_SDK_HEADER = 'x-cdp-sdk';
/** @hidden */
export const X_REQUEST_ID = 'X-Request-Id';
