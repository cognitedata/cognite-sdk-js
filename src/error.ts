// Copyright 2019 Cognite AS

import { AxiosError } from 'axios';

export class CogniteError extends Error {
  public status: number;
  public requestId?: string;
  public missing?: object[];
  public duplicated?: any[];
  /** @hidden */
  constructor(
    errorMessage: string,
    status: number,
    requestId?: string,
    extra: any = {}
  ) {
    let message = `${errorMessage} | code: ${status}`;
    if (requestId) {
      message += ` | X-Request-ID: ${requestId}`;
    }
    const { missing, duplicated } = extra;
    if (missing || duplicated) {
      message += `\n${JSON.stringify(extra, null, 2)}`;
    }
    super(message);
    Object.setPrototypeOf(this, CogniteError.prototype);
    this.status = status;
    this.requestId = requestId;
    this.missing = missing;
    this.duplicated = duplicated;
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(message).stack;
    }
  }
}

/** @hidden */
export function handleErrorResponse(err: AxiosError) {
  let code;
  let message;
  let requestId;
  let missing;
  let duplicated;
  try {
    code = err.response!.status;
    message = err.response!.data.error.message;
    missing = err.response!.data.error.missing;
    duplicated = err.response!.data.error.duplicated;
    requestId = (err.response!.headers || {})['X-Request-Id'];
  } catch (_) {
    throw err;
  }
  throw new CogniteError(message, code, requestId, { missing, duplicated });
}
