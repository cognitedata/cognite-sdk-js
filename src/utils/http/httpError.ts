// Copyright 2019 Cognite AS
import { HttpHeaders } from './basicHttpClient';

export class HttpError extends Error {
  /** @hidden */
  constructor(
    public status: number,
    public data: any,
    public headers: HttpHeaders
  ) {
    super(`Request failed | status code: ${status}`);
    Object.setPrototypeOf(this, HttpError.prototype); // https://stackoverflow.com/questions/51229574/why-instanceof-returns-false-for-a-child-object-in-javascript
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(this.message).stack;
    }
  }
}
