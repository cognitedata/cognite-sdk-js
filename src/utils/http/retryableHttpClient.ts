// Copyright 2019 Cognite AS
import { sleepPromise } from '../../utils';
import { BasicHttpClient, HttpRequest, HttpResponse } from './basicHttpClient';

export class RetryableHttpClient extends BasicHttpClient {
  private static calculateRetryDelayInMs(retryCount: number) {
    const INITIAL_RETRY_DELAY_IN_MS = 250;
    return (
      INITIAL_RETRY_DELAY_IN_MS + ((Math.pow(2, retryCount) - 1) / 2) * 1000
    );
  }

  constructor(baseUrl: string, private retryValidator: RetryValidator) {
    super(baseUrl);
  }

  protected async rawRequest<ResponseType>(
    request: HttpRequest
  ): Promise<HttpResponse<ResponseType>> {
    let retryCount = 0;
    do {
      const response = await super.rawRequest<ResponseType>(request);
      const shouldRetry = this.retryValidator(request, response, retryCount);
      if (!shouldRetry) {
        return response;
      }
      const delayInMs = RetryableHttpClient.calculateRetryDelayInMs(retryCount);
      await sleepPromise(delayInMs);
      retryCount++;
    } while (true);
  }
}

/** @hidden */
export type RetryValidator = (
  request: HttpRequest,
  response: HttpResponse<any>,
  retryCount: number
) => boolean;
