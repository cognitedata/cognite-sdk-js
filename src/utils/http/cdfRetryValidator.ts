// Copyright 2019 Cognite AS
import { inRange, some } from 'lodash';
import { HttpMethod } from './basicHttpClient';
import { RetryValidator } from './retryableHttpClient';

const MAX_RETRY_ATTEMPTS = 5;

const cdfEndpointsToRetry: { [key: string]: string[] } = {
  [HttpMethod.Post]: [
    '/assets/list',
    '/assets/byids',
    '/assets/search',
    '/events/list',
    '/events/byids',
    '/events/search',
    '/files/list',
    '/files/byids',
    '/files/search',
    '/files/initupload',
    '/files/downloadlink',
    '/timeseries/byids',
    '/timeseries/search',
    '/timeseries/data',
    '/timeseries/data/list',
    '/timeseries/data/latest',
    '/timeseries/data/delete',
  ],
};

export const cdfRetryValidator: RetryValidator = (
  request,
  response,
  retryCount
) => {
  if (retryCount >= MAX_RETRY_ATTEMPTS) {
    return false;
  }
  if (!isValidRetryStatusCode(response.status)) {
    return false;
  }
  if (universalRetryValidator(request, response, retryCount)) {
    return true;
  }
  const endpoints = cdfEndpointsToRetry[request.method] || [];
  if (matchPathWithEndpoints(request.path, endpoints)) {
    return true;
  }
  return false;
};

const universalRetryValidator: RetryValidator = (
  request,
  response,
  retryCount
) => {
  if (retryCount >= MAX_RETRY_ATTEMPTS) {
    return false;
  }
  const httpMethodsToRetry = ['GET', 'HEAD', 'OPTIONS', 'DELETE', 'PUT'];
  const isRetryableHttpMethod =
    httpMethodsToRetry.indexOf(request.method.toUpperCase()) !== -1;
  if (!isRetryableHttpMethod) {
    return false;
  }
  return isValidRetryStatusCode(response.status);
};

function isValidRetryStatusCode(status: number) {
  return (
    inRange(status, 100, 200) ||
    inRange(status, 429, 430) ||
    inRange(status, 500, 600)
  );
}

function matchPathWithEndpoints(path: string, endpoints: string[]) {
  return some(endpoints, endpoint => matchPathWithEndpoint(path, endpoint));
}

function matchPathWithEndpoint(path: string, endpoint: string) {
  return path.toUpperCase().indexOf(endpoint.toUpperCase()) !== -1;
}
