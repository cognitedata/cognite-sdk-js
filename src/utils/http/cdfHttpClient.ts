// Copyright 2019 Cognite AS

import * as Url from 'url';
import {
  API_KEY_HEADER,
  AUTHORIZATION_HEADER,
  X_CDF_APP_HEADER,
  X_CDF_SDK_HEADER,
} from '../../constants';
import { handleErrorResponse } from '../../error';
import {
  bearerString,
  isJson,
  transformDateInRequest,
  transformDateInResponse,
} from '../../utils';
import {
  HttpHeaders,
  HttpQueryParams,
  HttpRequest,
  HttpResponse,
} from './basicHttpClient';
import { cdfRetryValidator } from './cdfRetryValidator';
import { HttpError } from './httpError';
import { RetryableHttpClient } from './retryableHttpClient';

export class CDFHttpClient extends RetryableHttpClient {
  private static serializeQueryParameters(
    params: HttpQueryParams = {}
  ): HttpQueryParams {
    return Object.keys(params).reduce(
      (serializedParams, key) => {
        const param = params[key];
        serializedParams[key] = isJson(param) ? JSON.stringify(param) : param;
        return serializedParams;
      },
      {} as HttpQueryParams
    );
  }

  private static transformDateInResponse<T>(
    response: HttpResponse<T>
  ): HttpResponse<T> {
    const { data } = response;
    if (!isJson(data)) {
      return response;
    }
    return {
      ...response,
      data: transformDateInResponse(data),
    };
  }

  private static isSameOrigin(baseUrl: string, url: string) {
    const { protocol: baseUrlProtocol, host: baseUrlHost } = Url.parse(baseUrl);
    const { protocol, host } = Url.parse(Url.resolve(baseUrl, url));
    const hasSameProtocol = baseUrlProtocol === protocol;
    const hasSameHost = baseUrlHost === host;
    return hasSameProtocol && hasSameHost;
  }

  private static filterHeaders(headers: HttpHeaders, names: string[]) {
    return names.reduce(
      (partiallyFilteredHeaders, headerName) =>
        CDFHttpClient.filterHeader(partiallyFilteredHeaders, headerName),
      headers
    );
  }

  private static filterHeader(headers: HttpHeaders, name: string) {
    const filteredHeaders: HttpHeaders = {
      ...headers,
    };
    delete filteredHeaders[name];
    return filteredHeaders;
  }
  constructor(baseUrl: string) {
    super(baseUrl, cdfRetryValidator);
  }

  public setBearerToken(token: string) {
    this.setDefaultHeader(AUTHORIZATION_HEADER, bearerString(token));
  }

  public set401ResponseHandler(handler: Response401Handler) {
    this.response401Handler = handler;
  }

  protected async preRequest(request: HttpRequest): Promise<HttpRequest> {
    const headersWithDefaultHeaders = this.populateDefaultHeaders(
      request.headers
    );
    const headers = this.preventTokenLeakage(
      headersWithDefaultHeaders,
      request.path
    );
    const data = transformDateInRequest(request.data);
    const serializedQueryParams = CDFHttpClient.serializeQueryParameters(
      request.params
    );
    const params = transformDateInRequest(serializedQueryParams);
    return {
      ...request,
      data,
      headers,
      params,
    };
  }

  protected async postRequest<T>(
    response: HttpResponse<T>,
    request: HttpRequest
  ): Promise<HttpResponse<T>> {
    const transformedResponse = CDFHttpClient.transformDateInResponse(response);
    try {
      return await super.postRequest(transformedResponse, request);
    } catch (err) {
      if (err.status === 401 && !this.isLoginOrLogoutApi(request.path)) {
        return new Promise((resolvePromise, rejectPromise) => {
          const retry = () => resolvePromise(this.request(request));
          const reject = () => rejectPromise(err);
          this.response401Handler(err, retry, reject);
        });
      }
      throw handleErrorResponse(err);
    }
  }

  private response401Handler: Response401Handler = (_, __, reject) => reject();

  private preventTokenLeakage(headers: HttpHeaders, path: string) {
    if (CDFHttpClient.isSameOrigin(this.baseUrl, path)) {
      return headers;
    }
    return CDFHttpClient.filterHeaders(headers, [
      AUTHORIZATION_HEADER,
      API_KEY_HEADER,
      X_CDF_APP_HEADER,
      X_CDF_SDK_HEADER,
    ]);
  }

  private isLoginOrLogoutApi(url: string) {
    const lowerCaseUrl = url.toLowerCase();
    return (
      lowerCaseUrl.indexOf('/logout/url') !== -1 ||
      lowerCaseUrl.indexOf('/login/status') !== -1
    );
  }
}

type Response401Handler = (
  err: HttpError,
  retry: () => void,
  reject: () => void
) => void;
