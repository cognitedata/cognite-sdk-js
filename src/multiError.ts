// Copyright 2019 Cognite AS

import { CogniteError } from './error';
import { HttpResponse } from './utils/http/basicHttpClient';

/** @hidden */
export interface MultiErrorRawSummary<RequestType, ResponseType> {
  succeded: RequestType[][];
  responses: HttpResponse<ResponseType>[];
  failed: RequestType[][];
  errors: (Error | CogniteError)[];
}

function serialiseErrors(errors: (Error | CogniteError)[]) {
  return errors.map(err => {
    if (err instanceof CogniteError) {
      return err;
    }
    const { message, name, stack } = err;
    return { name, message, stack };
  });
}

export class CogniteMultiError<RequestType, ResponseType> extends Error {
  public failed: RequestType[];
  public succeded: RequestType[];

  public status?: number;
  public requestId?: string;
  public missing: object[] = [];
  public duplicated: any[] = [];
  public statuses: number[] = [];
  public requestIds: string[] = [];
  public responses: ResponseType[] = [];
  public errors: (Error | CogniteError)[];

  constructor({
    succeded,
    responses,
    failed,
    errors,
  }: MultiErrorRawSummary<RequestType, ResponseType>) {
    super(
      `The API Failed to process some items\n${JSON.stringify(
        {
          succeded,
          responses,
          failed,
          errors: serialiseErrors(errors),
        },
        null,
        2
      )}`
    );
    Object.setPrototypeOf(this, CogniteMultiError.prototype); // https://stackoverflow.com/questions/51229574/why-instanceof-returns-false-for-a-child-object-in-javascript

    this.responses = responses.map(response => response.data);
    this.errors = errors;
    this.succeded = succeded.reduce((all, current) => all.concat(current), []);
    this.failed = failed.reduce((all, current) => all.concat(current), []);
    this.status = (errors[0] as CogniteError).status;
    this.requestId = (errors[0] as CogniteError).requestId;

    for (const err of errors) {
      if (err instanceof CogniteError) {
        if (err.missing) {
          this.missing = this.missing.concat(err.missing);
        }
        if (err.duplicated) {
          this.duplicated = this.duplicated.concat(err.duplicated);
        }
        if (err.requestId) {
          this.requestIds.push(err.requestId);
        }
        this.statuses.push(err.status);
      }
    }

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(this.message).stack;
    }
  }
}
