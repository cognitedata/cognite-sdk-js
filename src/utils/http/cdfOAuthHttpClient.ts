// Copyright 2019 Cognite AS

import { AUTHORIZATION_HEADER } from '@/constants';
import { CogniteLoginError } from '@/loginError';
import {
  bearerString,
  createInvisibleIframe,
  generatePopupWindow,
  isSameProject,
  removeQueryParameterFromUrl,
} from '@/utils';
import { isString } from 'lodash';
import { parse, stringify } from 'query-string';
import { CDFHttpClient } from './cdfHttpClient';
import { HttpError } from './httpError';

export class CDFOAuthHttpClient extends CDFHttpClient {
  private static loginWithRedirect(params: AuthorizeParams): Promise<void> {
    const unresolveablePromise = new Promise<void>(() => {
      const url = CDFOAuthHttpClient.generateLoginUrl(params);
      window.location.assign(url);
    });
    return unresolveablePromise;
  }

  private static loginWithPopup(
    params: AuthorizeParams
  ): Promise<null | AuthTokens> {
    return new Promise((resolve, reject) => {
      const url = CDFOAuthHttpClient.generateLoginUrl(params);
      const loginPopupWindow = generatePopupWindow(url, LOGIN_POPUP_NAME);
      if (loginPopupWindow === null) {
        reject(new CogniteLoginError('Failed to create login popup window'));
        return;
      }
      const tokenListener = (message: MessageEvent) => {
        const isUntrustedSource = message.source !== loginPopupWindow;
        if (isUntrustedSource) {
          return;
        }
        if (!message.data || message.data.error) {
          reject(new CogniteLoginError());
          return;
        }
        window.removeEventListener('message', tokenListener);
        resolve(message.data);
      };
      window.addEventListener('message', tokenListener, false);
    });
  }

  private static silentLogin(params: AuthorizeParams): Promise<AuthTokens> {
    return new Promise<AuthTokens>((resolve, reject) => {
      const url = `${CDFOAuthHttpClient.generateLoginUrl(params)}&prompt=none`;
      const iframe = createInvisibleIframe(url, LOGIN_IFRAME_NAME);
      iframe.onload = () => {
        try {
          const authTokens = CDFOAuthHttpClient.parseTokenQueryParameters(
            iframe.contentWindow!.location.search
          );
          if (authTokens === null) {
            throw Error('Failed to login');
          }
          resolve(authTokens);
        } catch (e) {
          reject(e);
        } finally {
          document.body.removeChild(iframe);
        }
      };
      document.body.appendChild(iframe);
    });
  }

  private static generateLoginUrl({
    project,
    baseUrl,
    redirectUrl,
    errorRedirectUrl,
  }: AuthorizeParams): string {
    const queryParams = {
      project,
      redirectUrl,
      errorRedirectUrl: errorRedirectUrl || redirectUrl,
    };
    return `${baseUrl}/login/redirect?${stringify(queryParams)}`;
  }

  private static extractTokensFromUrl() {
    try {
      const tokens = CDFOAuthHttpClient.parseTokenQueryParameters(
        window.location.search
      );
      CDFOAuthHttpClient.removeQueryParametersFromUrl([
        ACCESS_TOKEN_PARAM,
        ID_TOKEN_PARAM,
      ]);
      return tokens;
    } catch (err) {
      CDFOAuthHttpClient.removeQueryParametersFromUrl([
        ERROR_PARAM,
        ERROR_DESCRIPTION_PARAM,
      ]);
      throw err;
    }
  }

  private static parseTokenQueryParameters(query: string): null | AuthTokens {
    const {
      [ACCESS_TOKEN_PARAM]: accessToken,
      [ID_TOKEN_PARAM]: idToken,
      [ERROR_PARAM]: error,
      [ERROR_DESCRIPTION_PARAM]: errorDescription,
    } = parse(query);
    if (error !== undefined) {
      throw Error(`${error}: ${errorDescription}`);
    }
    if (isString(accessToken) && isString(idToken)) {
      return {
        accessToken,
        idToken,
      };
    }
    return null;
  }

  private static removeQueryParametersFromUrl(params: string[]): void {
    let url = window.location.href;
    params.forEach(param => {
      url = removeQueryParameterFromUrl(url, param);
    });
    window.history.replaceState(null, '', url);
  }

  private static isLoginPopupWindow(): boolean {
    return window.name === LOGIN_POPUP_NAME;
  }

  private static isAuthIFrame(): boolean {
    return window.name === LOGIN_IFRAME_NAME;
  }

  private static async loginSilently(
    params: AuthenticateParams
  ): Promise<null | AuthTokens> {
    const { baseUrl, project } = params;
    const href = window.location.href;
    const tokens = await CDFOAuthHttpClient.silentLogin({
      baseUrl,
      project,
      redirectUrl: href,
      errorRedirectUrl: href,
    });
    return tokens;
  }

  constructor(baseUrl: string) {
    super(baseUrl);
  }

  private async validateTokens(project: string, tokens: AuthTokens) {
    const idInfo = await super.getIdInfo({
      [AUTHORIZATION_HEADER]: bearerString(tokens.accessToken),
    });
    return idInfo !== null && isSameProject(idInfo.project, project);
  }
}

/** @hidden */
export interface AuthorizeParams extends AuthorizeOptions {
  baseUrl: string;
  project: string;
  accessToken?: string;
}

export interface AuthorizeOptions {
  redirectUrl: string;
  errorRedirectUrl?: string;
}

/** @hidden */
export interface AuthenticateParams {
  project: string;
  baseUrl: string;
}

export interface AuthTokens {
  accessToken: string;
  idToken: string;
}

const ACCESS_TOKEN_PARAM = 'access_token';
const ID_TOKEN_PARAM = 'id_token';
const ERROR_PARAM = 'error';
const ERROR_DESCRIPTION_PARAM = 'error_description';
const LOGIN_IFRAME_NAME = 'cognite-js-sdk-auth-iframe';
const LOGIN_POPUP_NAME = 'cognite-js-sdk-auth-popup';

type NotAuthenticatedHandler = (
  err: HttpError,
  retry: () => void,
  ignore: () => void
) => void;
