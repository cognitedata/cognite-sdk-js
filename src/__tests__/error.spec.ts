// Copyright 2019 Cognite AS

import { X_REQUEST_ID } from '../constants';
import { CogniteError, handleErrorResponse } from '../error';
import { CogniteMultiError } from '../multiError';
import { HttpError } from '../utils/http/httpError';
import { createErrorReponse } from './testUtils';

const internalIdObject = { id: 4190022127342195 };
const externalIdObject = { externalId: 'abc' };
const event = {
  externalId: 'string',
  startTime: 0,
  endTime: 0,
  description: 'string',
  source: 'string',
};

describe('CogniteError', () => {
  test('without requestId', () => {
    const errorMessage = 'Abc';
    const status = 500;
    const error = new CogniteError(errorMessage, status);
    expect(error.status).toBe(status);
    expect(error.requestId).toBeUndefined();
    expect(() => {
      throw error;
    }).toThrowErrorMatchingInlineSnapshot(`"Abc | code: 500"`);
  });

  test('with requestId', () => {
    const errorMessage = 'Abc';
    const status = 500;
    const requestId = 'def';
    const error = new CogniteError(errorMessage, status, requestId);
    expect(error.status).toBe(status);
    expect(error.requestId).toBe(requestId);
    expect(() => {
      throw error;
    }).toThrowErrorMatchingInlineSnapshot(
      `"Abc | code: 500 | X-Request-ID: def"`
    );
  });

  test('extra field', () => {
    const errorMessage = 'Abc';
    const status = 500;
    const error = new CogniteError(errorMessage, status, undefined, {
      missing: [internalIdObject, externalIdObject],
      duplicated: [event],
    });
    expect(() => {
      throw error;
    }).toThrowErrorMatchingSnapshot();
  });
});

describe('handleErrorResponse', () => {
  test('without requestId', () => {
    const httpError = new HttpError(500, createErrorReponse(500, 'abc'), {});
    expect(() => {
      handleErrorResponse(httpError);
    }).toThrowErrorMatchingInlineSnapshot(`"abc | code: 500"`);
  });

  test('with requestId', () => {
    const httpError = new HttpError(500, createErrorReponse(500, 'abc'), {
      [X_REQUEST_ID]: 'def',
    });
    expect(() => {
      handleErrorResponse(httpError);
    }).toThrowErrorMatchingInlineSnapshot(
      `"abc | code: 500 | X-Request-ID: def"`
    );
  });

  test('extra fields', () => {
    const status = 500;
    const message = 'abc';
    const xRequestId = 'def';
    const httpError = new HttpError(
      status,
      createErrorReponse(status, message, {
        missing: [internalIdObject, externalIdObject],
        duplicated: [event],
      }),
      { [X_REQUEST_ID]: xRequestId }
    );

    expect(() => {
      handleErrorResponse(httpError);
    }).toThrowErrorMatchingSnapshot();

    try {
      handleErrorResponse(httpError);
    } catch (e) {
      expect(e.status).toBe(status);
      expect(e.requestId).toBe(xRequestId);
      expect(e.missing).toEqual([internalIdObject, externalIdObject]);
      expect(e.duplicated).toEqual([event]);
    }
  });
});

describe('Cognite multi error', () => {
  test('create with 2 fails and 1 success', () => {
    const errMsg = 'createAssets.arg0.items: size must be between 1 and 1000';
    const nestedErr = new CogniteError(errMsg, 400, 'r1', {
      missing: ['something'],
    });
    const nestedErr2 = new CogniteError(errMsg, 500, 'r2', {
      missing: ['more'],
      duplicated: ['this one'],
    });
    const err = new CogniteMultiError({
      succeded: [[2]],
      failed: [[0, 1]],
      errors: [nestedErr, nestedErr2],
      responses: [],
    });

    expect(err.succeded).toEqual([2]);
    expect(err.failed).toEqual([0, 1]);
    expect(err.statuses).toEqual([400, 500]);
    expect(err.status).toEqual(400);
    expect(err.requestId).toEqual('r1');
    expect(err.errors).toEqual([nestedErr, nestedErr2]);
    expect(err.message).toMatchSnapshot();
    expect(err.missing).toEqual(['something', 'more']);
    expect(err.duplicated).toEqual(['this one']);
    expect(err.requestIds).toEqual(['r1', 'r2']);
  });

  test('multierror serialises non-api errors', () => {
    const unknownError = new Error('unknown');

    const err = new CogniteMultiError({
      failed: [],
      errors: [unknownError],
      succeded: [],
      responses: [],
    });

    expect(err.message).toContain(`"message": "${unknownError.message}"`);
  });
});
