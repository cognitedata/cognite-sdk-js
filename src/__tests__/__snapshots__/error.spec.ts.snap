// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cognite multi error create with 2 fails and 1 success 1`] = `
"The API Failed to process some items
{
  \\"succeded\\": [
    [
      2
    ]
  ],
  \\"responses\\": [],
  \\"failed\\": [
    [
      0,
      1
    ]
  ],
  \\"errors\\": [
    {
      \\"status\\": 400,
      \\"message\\": \\"createAssets.arg0.items: size must be between 1 and 1000\\",
      \\"requestId\\": \\"r1\\",
      \\"missing\\": [
        \\"something\\"
      ]
    },
    {
      \\"status\\": 500,
      \\"message\\": \\"createAssets.arg0.items: size must be between 1 and 1000\\",
      \\"requestId\\": \\"r2\\",
      \\"missing\\": [
        \\"more\\"
      ],
      \\"duplicated\\": [
        \\"this one\\"
      ]
    }
  ]
}"
`;

exports[`CogniteError extra field 1`] = `
"Abc | code: 500
{
  \\"missing\\": [
    {
      \\"id\\": 4190022127342195
    },
    {
      \\"externalId\\": \\"abc\\"
    }
  ],
  \\"duplicated\\": [
    {
      \\"externalId\\": \\"string\\",
      \\"startTime\\": 0,
      \\"endTime\\": 0,
      \\"description\\": \\"string\\",
      \\"source\\": \\"string\\"
    }
  ]
}"
`;

exports[`handleErrorResponse extra fields 1`] = `
"abc | code: 500 | X-Request-ID: def
{
  \\"missing\\": [
    {
      \\"id\\": 4190022127342195
    },
    {
      \\"externalId\\": \\"abc\\"
    }
  ],
  \\"duplicated\\": [
    {
      \\"externalId\\": \\"string\\",
      \\"startTime\\": 0,
      \\"endTime\\": 0,
      \\"description\\": \\"string\\",
      \\"source\\": \\"string\\"
    }
  ]
}"
`;
